### PR TITLE
Trigger token through the webui

### PR DESCRIPTION
**Pull request by krauselukas**
Enable users to trigger their release, rebuild and service tokens through the webui, similar to what is already possible through the API. PR includes a couple of refactorings in order to reuse existing code.

### CODE REVIEW

File: src/api/app/controllers/webui/users/tokens_controller.rb [hunk 1]
```diff
@@ -57,6 +59,25 @@ def destroy
     redirect_to tokens_url
   end
 
+  # method for the view
```
Comment (vpereira) [inline src/api/app/controllers/webui/users/tokens_controller.rb:62]: not possible to move it to a new controller, subcontroller from TokensController with just some CRUD methods, instead such custom method names?

Comment (krauselukas) [reply src/api/app/controllers/webui/users/tokens_controller.rb:62]: Already on it :)

File: src/api/app/views/webui/users/token_triggers/show.html.haml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+              .form-group
+                = token_fields.label(:package, 'Package:')
+                = token_package.name
+          - else
+            .form-group.row
+              .col-sm
+                = render partial: 'webui/shared/autocomplete', locals: { html_id: 'project_name', label: 'Project:', required: true,
+                                                                         data: { source: autocomplete_projects_path(local: true) } }
+              .col-sm
+                = render partial: 'webui/shared/autocomplete', locals: { html_id: 'package_name', label: 'Package:', required: true,
```
Comment (hennevogel) [inline src/api/app/views/webui/users/token_triggers/show.html.haml:32]: This seems broken if results is a single package. Just tried in the review app....

Screenshot from 2021-09-09 13-54-32

Screenshot from 2021-09-09 13-54-55

Comment (krauselukas) [reply src/api/app/views/webui/users/token_triggers/show.html.haml:32]: alright, will have a look!

File: src/api/spec/controllers/trigger_controller_spec.rb [hunk 3]
```diff
@@ -17,6 +15,12 @@
     package
   end
 
+  it 'includes Triggerable concern' do
```
Comment (dmarcoux) [inline src/api/spec/controllers/trigger_controller_spec.rb:18]: I don't think this spec brings any value. I would remove it.

File: src/api/spec/controllers/concerns/triggerable_spec.rb [hunk 4]
```diff
@@ -0,0 +1,173 @@
+require 'rails_helper'
+
+RSpec.describe Triggerable do
+  let(:fakes_controller) do
```
Comment (vpereira) [inline src/api/spec/controllers/concerns/triggerable_spec.rb:4]: `fake_controller` without s

File: src/api/spec/controllers/concerns/triggerable_spec.rb [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    Class.new(ApplicationController) do
+      include Triggerable
+    end
+  end
+
+  let(:object) { fakes_controller.new }
```
Comment (vpereira) [inline src/api/spec/controllers/concerns/triggerable_spec.rb:10]: `fake_controller_instance` is better than `object`

File: src/api/spec/controllers/concerns/triggerable_spec.rb [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+  let(:user) { create(:confirmed_user, login: 'foo') }
+  let(:project) { create(:project, name: 'project', maintainer: user) }
+  let(:package) { create(:package, name: 'package_trigger', project: project) }
+  let(:repository) { create(:repository, name: 'package_test_repository', architectures: ['x86_64'], project: project) }
+
+  before do
+    object.instance_variable_set(:@token, token)
```
Comment (vpereira) [inline src/api/spec/controllers/concerns/triggerable_spec.rb:18]: if you have a method `token` (even being private) on the object, you can work with `allow` and `receive`  too

File: src/api/spec/controllers/concerns/triggerable_spec.rb [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+      it 'raises when package does not exist in link' do
+        stub_params(project_name: project_with_a_link.name, package_name: 'does-not-exist')
+        object.set_project
+
+        expect { object.set_package }.to raise_error(Package::Errors::UnknownObjectError)
+      end
+
+      it 'assigns linked package' do
+        stub_params(project_name: project_with_a_link.name, package_name: package.name)
+        object.set_project
```
Comment (vpereira) [inline src/api/spec/controllers/concerns/triggerable_spec.rb:61]: unless you are testing he methods `set_project`, `set_package` and `set_object_to_authorize`, you could as well stub them with `allow` and `receive`

File: src/api/spec/controllers/trigger_controller_spec.rb (hunk 3 above)
Comment (krauselukas) [reply src/api/spec/controllers/trigger_controller_spec.rb]: done

File: src/api/spec/controllers/concerns/triggerable_spec.rb (hunk 4 above)
Comment (krauselukas) [reply src/api/spec/controllers/concerns/triggerable_spec.rb]: done

File: src/api/spec/controllers/concerns/triggerable_spec.rb (hunk 5 above)
Comment (krauselukas) [reply src/api/spec/controllers/concerns/triggerable_spec.rb]: done

File: src/api/spec/controllers/concerns/triggerable_spec.rb (hunk 6 above)
Comment (krauselukas) [reply src/api/spec/controllers/concerns/triggerable_spec.rb]: but `token` is an instance variable, with the `allow` and `receive` it does not seem to be possible to mock instance variables

**Review by saraycp — APPROVED** [review]
LGTM

**Comment by vpereira** [comment]
@krauselukas please change the PR description :)

**Comment by krauselukas** [comment]
@vpereira PR description updated

**Pull request merged**